clone/cmd/ctclone/ctclone.go

```diff
@@ -0,0 +1,178 @@
+	writeBatchSize = flag.Uint("write_batch_size", 32, "The number of leaves to write in each DB transaction.")
+	workers        = flag.Uint("workers", 2, "The number of worker threads to run in parallel to fetch entries.")
+)
+
+func main() {
+	flag.Parse()
+
+	if !strings.HasSuffix(*logURL, "/") {
+		glog.Exit("'log_url' must end with '/'")
+	}
+	if len(*mysqlURI) == 0 {
+		glog.Exit("Missing required parameter 'mysql_uri'")
+	}
+	lu, err := url.Parse(*logURL)
+	if err != nil {
+		glog.Exitf("log_url is invalid: %v", err)
+	}
+
+	ctx := context.Background()
+	db, err := logdb.NewDatabase(*mysqlURI)
+	if err != nil {
+		glog.Exitf("Failed to connect to database: %q", err)
+	}
+	// TODO(mhutchinson): Persist log URI and check here to prevent user-error writing different logs to same DB.
+	head, err := db.Head()
+	if err != nil {
+		if err == logdb.ErrNoDataFound {
+			glog.Infof("failed to find head of database, assuming empty and starting from scratch: %v", err)
+			head = -1
+		} else {
+			glog.Exitf("Failed to query for head of local log: %q", err)
+		}
+	}
+	next := uint64(head + 1)
+
+	fetcher := download.NewHTTPFetcher(lu)
+
+	errChan := make(chan error)
+	lc := make(chan []byte, *writeBatchSize*2)
+	go download.Bulk(next, certLeafFetcher{fetcher}.Batch, *workers, *fetchBatchSize, lc, errChan)
+
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	r := reporter{
+		lastReported: next,
+		lastReport:   time.Now(),
+	}
+
+	batch := make([][]byte, *writeBatchSize)
+	bi := 0
+	bs := next
+
+	for {
+		select {
+		case err := <-errChan:
+			glog.Exit(err)
+		case <-ticker.C:
+			r.report()
+		case l := <-lc:
+			workDone := r.trackWork(next)
+			next++
+			batch[bi] = l
+			bi = (bi + 1) % int(*writeBatchSize)
+			if bi == 0 {
+				if err := db.WriteLeaves(ctx, bs, batch); err != nil {
+					glog.Exitf("Failed to write to DB for batch starting at %d: %q", bs, err)
+				}
+				bs = next
+			}
+			workDone()
+		}
+	}
+}
+
+type reporter struct {
+	lastReport   time.Time
+	lastReported uint64
+	lastWorked   uint64
+	epochWorked  time.Duration
+}
+
+func (r *reporter) report() {
+	elapsed := time.Since(r.lastReport)
+	workRatio := r.epochWorked.Seconds() / elapsed.Seconds()
+
+	rate := float64(r.lastWorked-r.lastReported) / elapsed.Seconds()
+	glog.Infof("%.1f leaves/s, last leaf=%d, time working=%.1f%%", rate, r.lastReported, 100*workRatio)
+	r.epochWorked = 0
+	r.lastReported = r.lastWorked
+	r.lastReport = time.Now()
+}
+
+func (r *reporter) trackWork(index uint64) func() {
+	start := time.Now()
+	r.lastWorked = index
+
+	return func() {
+		end := time.Now()
+		r.epochWorked += end.Sub(start)
+	}
+}
+
+// fetcher gets data paths. This allows impl to be swapped for tests.
+type fetcher interface {
+	// GetData gets the data at the given path, or returns an error.
+	GetData(path string) ([]byte, error)
+}
+
+type certLeafFetcher struct {
+	f fetcher
+}
+
+func (clf certLeafFetcher) Batch(start uint64, leaves [][]byte) error {
```

Comment:
Having `leaves` as an OUT parameter is a bit weird - caused me a double-take on what this function does when reading the signature.

I assume the reason you're doing this is to try to help the garbage collector/allocator out? 
(If so, did you benchmark to see whether it helps?)

If you're set on keeping it this way, maybe consider changing the signature to something like `... Populate(start, batch [][]byte)` - though having written that out `start` is a bit ambiguous (e.g. "is it the first index in `batch` to fill from?")